px-context-browser-item.html

```diff
@@ -50,7 +98,120 @@
+     * element and returns an object with the details.
+     *
+     * @param {HTMLElement} el
+     * @returns {Object} styles
+     * @returns {String} styles.size - The font size as a string with px suffix (e.g. '12px')
+     * @returns {String} styles.family - The CSS font family declaration as a string (e.g. 'Helvetica')
+     */
+    getFontStyles(el) {
+        var styles = window.getComputedStyle(el);
+        var size = styles.fontSize;
+        var family = styles.fontFamily;
+        return {
+            size: size,
+            family: family
+        }
+    },
+
+    /**
+     * Iterates over a list of items and determines which will fit into the
+     * available width and which will not.
+     *
+     * If any items do not fit, backtracks and reduces the available width to
+     * ensure an overflowed icon will also fit.
+     *
+     * @param  {Array} items - A flat Array of items to fit
+     * @param {Object} fontStyles - An object with the font size and family strings from CSSText
+     * @param  {Number} width - The width to fit the items into
+     * @return {Array.<Array>} - First entry is Array of items that fit, second entry is Array of items that do not fit
+     */
+    truncateBreadcrumbs(items, fontStyles, width) {
+      if (!items.length) {
+        return ''; // no breadcrumbs to display for the root item
       }
+      const ellipWidth = this.getTextWidth("\u2026 > ", fontStyles);
+      const caratWidth = this.getTextWidth(" > ", fontStyles);
+
+      const available = width;
+      const len = items.length;
+      let visible = len;
+      let lastResult;
+
+      while (visible > 0) {
+        // Get array items from end to beginning
+        let itemsToFit = items.slice(visible * -1);
+        // Join items together with carat
+        let textToMeasure = itemsToFit.join(' > ');
+        // If some items fit but not all, prepend ellipses
+        if (visible < len && len > 1) {
+          textToMeasure =  '... > ' + textToMeasure;
+        }
+        // Measure the resulting string to see if it fits
+        let measure = this.getTextWidth(textToMeasure, fontStyles);
+        // If it fits, we're done!
+        if (available > measure) {
+          return textToMeasure;
+        }
+        // Otherwise, try to fit with one less item
+        visible--;
+      }
+
+      // We know visible === 0, truncate last item and add ... > if needed
+      let prefix = (len > 1) ? '... > ' : '';
+      // CSS will take care of any end truncation needed
+      return prefix + items[len-1];
+    },
+
+    /**
+     * Measures the given text and returns the width based on font size and family
+     *
+     * @param  {String} text - A text string
+     * @param {Object} fontStyles - An object with the font size and family strings from CSSText
+     * @return {Number} - The rounded width of the text string
+     */
+    getTextWidth(text, fontStyles) {
+      var canvas = document.createElement('canvas');
```

Comment:
Wonder if we should cache and re-use the canvas element so we don't need to make a new one every time we need to measure text width.

Possible approach:

```
var canvas = this._textMeasureCanvas ? document.createElement('canvas') : this._textMeasureCanvas;
```

Reply:
`var canvas = this._textMeasureCanvas = (this._textMeasureCanvas ? document.createElement('canvas') : this._textMeasureCanvas);`

Reply:
👍